File: concepts/02-types-and-structs/go/store.go
// Package tasks holds Stage 2 exercises: zero values, and the nil-map vs
// nil-slice footgun that has no Dart analogue (Dart's Map/List are never
// "unset," just empty).
package tasks

// Task is a plain struct — its zero value (Task{}) is already fully usable,
// unlike a Dart class with late/nullable fields that must be assigned before
// use.
type Task struct {
	ID    string
	Title string
	Done  bool
}

// Store holds tasks in memory, keyed by ID.
//
// TODO(exercise, Level 2): Store's zero value has a nil map (tasks is
// unexported and never initialized here). Reading from a nil map is safe
// in Go — it just returns the zero value — but WRITING to a nil map
// panics. Fix Add so a zero-value Store (var s Store, or &Store{}) works
// without requiring a separate constructor call first.
type Store struct {
	tasks map[string]Task
}

// Add stores t, keyed by t.ID.
func (s *Store) Add(t Task) {
	// TODO: guard against s.tasks being nil (lazily initialize it here
	// with make(map[string]Task) before the first write), then:
	// s.tasks[t.ID] = t
}

// Get returns the task with the given id, and whether it was found.
// Already correct: reading a nil map is safe and returns (Task{}, false).
func (s *Store) Get(id string) (Task, bool) {
	t, ok := s.tasks[id]
	return t, ok
}

// Titles returns the title of every task, in no particular order.
// Already correct — included to contrast with Store.tasks: appending to a
// nil slice is always safe in Go, unlike writing to a nil map.
func (s *Store) Titles() []string {
	var titles []string // nil slice, not []string{} — and that's fine
	for _, t := range s.tasks {
		titles = append(titles, t.Title) // append on nil works
	}
	return titles
}

File: concepts/02-types-and-structs/dart/store.dart
/// Dart equivalent of go/store.go's Store.
/// There is no footgun to demonstrate here: Dart's Map and List fields are
/// never "unset" the way a Go zero-value map is nil — either they're
/// initialized in the constructor (as below) or the field is nullable and
/// you're forced to null-check it. Go gives you a THIRD state Dart doesn't
/// have: a non-nil, non-empty-literal, perfectly typed map that still
/// panics on write until you initialize it.
class Task {
  final String id;
  final String title;
  final bool done;

  Task(this.id, this.title, {this.done = false});
}

class Store {
  final Map<String, Task> _tasks = {}; // always initialized, never null

  void add(Task t) => _tasks[t.id] = t;

  Task? get(String id) => _tasks[id];

  List<String> titles() => _tasks.values.map((t) => t.title).toList();
}
